File: examples/Algorithm/python/19_greedy.py
"""
Greedy Algorithm
Greedy Algorithms

An algorithmic technique that makes the locally optimal choice at each step to find the global optimum.
"""

from typing import List, Tuple
import heapq


# =============================================================================
# 1. Activity Selection Problem
# =============================================================================

def activity_selection(activities: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
    """
    Select the maximum number of non-overlapping activities
    activities: [(start, end), ...]
    Greedy strategy: select activities with earliest end time first
    Time Complexity: O(n log n)
    """
    # Sort by end time
    sorted_activities = sorted(activities, key=lambda x: x[1])

    selected = []
    last_end = 0

    for start, end in sorted_activities:
        if start >= last_end:
            selected.append((start, end))
            last_end = end

    return selected


# =============================================================================
# 2. Fractional Knapsack Problem
# =============================================================================

def fractional_knapsack(capacity: int, items: List[Tuple[int, int]]) -> float:
    """
    Fractional knapsack problem
    items: [(value, weight), ...]
    Greedy strategy: select items with highest value-to-weight ratio first
    Time Complexity: O(n log n)
    """
    # Sort by value-to-weight ratio (descending)
    sorted_items = sorted(items, key=lambda x: x[0] / x[1], reverse=True)

    total_value = 0
    remaining = capacity

    for value, weight in sorted_items:
        if remaining >= weight:
            total_value += value
            remaining -= weight
        else:
            # Take a fraction
            total_value += value * (remaining / weight)
            break

    return total_value


# =============================================================================
# 3. Meeting Rooms
# =============================================================================

def min_meeting_rooms(intervals: List[Tuple[int, int]]) -> int:
    """
    Minimum number of meeting rooms required
    Time Complexity: O(n log n)
    """
    if not intervals:
        return 0

    # Separate into start/end events
    events = []
    for start, end in intervals:
        events.append((start, 1))   # Start: +1
        events.append((end, -1))    # End: -1

    events.sort()

    max_rooms = current_rooms = 0
    for _, delta in events:
        current_rooms += delta
        max_rooms = max(max_rooms, current_rooms)

    return max_rooms


# =============================================================================
# 4. Job Scheduling
# =============================================================================

def job_scheduling(jobs: List[Tuple[int, int, int]]) -> Tuple[int, List[int]]:
    """
    Schedule jobs for maximum profit within deadlines
    jobs: [(job_id, deadline, profit), ...]
    Greedy strategy: sort by profit descending, assign to latest available slot
    """
    # Sort by profit descending
    sorted_jobs = sorted(jobs, key=lambda x: x[2], reverse=True)

    max_deadline = max(job[1] for job in jobs)
    slots = [None] * (max_deadline + 1)  # Time slots
    total_profit = 0
    scheduled = []

    for job_id, deadline, profit in sorted_jobs:
        # Find an empty slot from deadline backwards
        for slot in range(deadline, 0, -1):
            if slots[slot] is None:
                slots[slot] = job_id
                total_profit += profit
                scheduled.append(job_id)
                break

    return total_profit, scheduled


# =============================================================================
# 5. Huffman Coding
# =============================================================================

class HuffmanNode:
    def __init__(self, char: str = None, freq: int = 0):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.freq < other.freq


def huffman_encoding(text: str) -> Tuple[dict, str]:
    """
    Huffman encoding
    Returns: (character->code dictionary, encoded string)
    Time Complexity: O(n log n)
    """
    if not text:
        return {}, ""

    # Calculate frequencies
    freq = {}
    for char in text:
        freq[char] = freq.get(char, 0) + 1

    # Build tree using priority queue
    heap = [HuffmanNode(char, f) for char, f in freq.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)

        merged = HuffmanNode(freq=left.freq + right.freq)
        merged.left = left
        merged.right = right

        heapq.heappush(heap, merged)

    # Generate codes
    codes = {}

    def generate_codes(node: HuffmanNode, code: str):
        if node is None:
            return

        if node.char is not None:
            codes[node.char] = code if code else "0"
            return

        generate_codes(node.left, code + "0")
        generate_codes(node.right, code + "1")

    if heap:
        generate_codes(heap[0], "")

    # Encode
    encoded = ''.join(codes[char] for char in text)

    return codes, encoded


def huffman_decoding(encoded: str, codes: dict) -> str:
    """Huffman decoding"""
    reverse_codes = {v: k for k, v in codes.items()}
    decoded = []
    current = ""

    for bit in encoded:
        current += bit
        if current in reverse_codes:
            decoded.append(reverse_codes[current])
            current = ""

    return ''.join(decoded)


# =============================================================================
# 6. Coin Change (Greedy)
# =============================================================================

def coin_change_greedy(coins: List[int], amount: int) -> List[int]:
    """
    Coin change (greedy, optimal only for certain coin sets)
    coins: coins sorted in descending order
    Note: DP is required for general coin combinations
    """
    coins = sorted(coins, reverse=True)
    result = []

    for coin in coins:
        while amount >= coin:
            result.append(coin)
            amount -= coin

    return result if amount == 0 else []


# =============================================================================
# 7. Interval Scheduling (Partitioning)
# =============================================================================

def interval_partitioning(intervals: List[Tuple[int, int]]) -> List[List[Tuple[int, int]]]:
    """
    Partition intervals into minimum number of groups (no overlap within each group)
    Time Complexity: O(n log n)
    """
    if not intervals:
        return []

    # Sort by start time
    sorted_intervals = sorted(enumerate(intervals), key=lambda x: x[1][0])

    # Manage end times of each group using a heap
    groups = []  # [(end_time, group_index)]
    assignment = [[] for _ in range(len(intervals))]

    for idx, (start, end) in sorted_intervals:
        if groups and groups[0][0] <= start:
            # Assign to existing group
            _, group_idx = heapq.heappop(groups)
            assignment[group_idx].append(intervals[idx])
            heapq.heappush(groups, (end, group_idx))
        else:
            # Create new group
            new_group = len(groups)
            assignment.append([intervals[idx]])
            heapq.heappush(groups, (end, new_group))

    return [g for g in assignment if g]


# =============================================================================
# 8. Jump Game
# =============================================================================

def can_jump(nums: List[int]) -> bool:
    """
    Whether the last index is reachable
    nums[i] = maximum jump distance from i
    Time Complexity: O(n)
    """
    max_reach = 0

    for i, jump in enumerate(nums):
        if i > max_reach:
            return False
        max_reach = max(max_reach, i + jump)

    return True


def min_jumps(nums: List[int]) -> int:
    """
    Minimum number of jumps to reach the last index
    Time Complexity: O(n)
    """
    if len(nums) <= 1:
        return 0

    jumps = 0
    current_end = 0
    farthest = 0

    for i in range(len(nums) - 1):
        farthest = max(farthest, i + nums[i])

        if i == current_end:
            jumps += 1
            current_end = farthest

            if current_end >= len(nums) - 1:
                break

    return jumps


# =============================================================================
# 9. Gas Station
# =============================================================================

def can_complete_circuit(gas: List[int], cost: List[int]) -> int:
    """
    Find starting station to complete a circular route
    gas[i] = fuel gained at station i
    cost[i] = fuel cost to travel from i to i+1
    Time Complexity: O(n)
    """
    total_tank = 0
    current_tank = 0
    start = 0

    for i in range(len(gas)):
        gain = gas[i] - cost[i]
        total_tank += gain
        current_tank += gain

        if current_tank < 0:
            start = i + 1
            current_tank = 0

    return start if total_tank >= 0 else -1


# =============================================================================
# 10. Minimum Arrows to Burst Balloons
# =============================================================================

def min_arrows(points: List[List[int]]) -> int:
    """
    Minimum number of arrows to burst all horizontal balloons
    points[i] = [start, end] range
    Time Complexity: O(n log n)
    """
    if not points:
        return 0

    # Sort by end point
    points.sort(key=lambda x: x[1])

    arrows = 1
    arrow_pos = points[0][1]

    for start, end in points[1:]:
        if start > arrow_pos:
            arrows += 1
            arrow_pos = end

    return arrows


# =============================================================================
# Tests
# =============================================================================

def main():
    print("=" * 60)
    print("Greedy Algorithm Examples")
    print("=" * 60)

    # 1. Activity Selection
    print("\n[1] Activity Selection Problem")
    activities = [(1, 4), (3, 5), (0, 6), (5, 7), (3, 9), (5, 9), (6, 10), (8, 11)]
    selected = activity_selection(activities)
    print(f"    Activities: {activities}")
    print(f"    Selected: {selected} ({len(selected)} activities)")

    # 2. Fractional Knapsack
    print("\n[2] Fractional Knapsack")
    items = [(60, 10), (100, 20), (120, 30)]  # (value, weight)
    capacity = 50
    value = fractional_knapsack(capacity, items)
    print(f"    Items (value, weight): {items}")
    print(f"    Capacity: {capacity}, Max value: {value}")

    # 3. Meeting Rooms
    print("\n[3] Minimum Meeting Rooms")
    meetings = [(0, 30), (5, 10), (15, 20)]
    rooms = min_meeting_rooms(meetings)
    print(f"    Meetings: {meetings}")
    print(f"    Rooms needed: {rooms}")

    # 4. Job Scheduling
    print("\n[4] Job Scheduling")
    jobs = [(1, 4, 20), (2, 1, 10), (3, 1, 40), (4, 1, 30)]  # (ID, deadline, profit)
    profit, scheduled = job_scheduling(jobs)
    print(f"    Jobs (ID, deadline, profit): {jobs}")
    print(f"    Schedule: {scheduled}, Total profit: {profit}")

    # 5. Huffman Coding
    print("\n[5] Huffman Coding")
    text = "abracadabra"
    codes, encoded = huffman_encoding(text)
    decoded = huffman_decoding(encoded, codes)
    print(f"    Original: '{text}'")
    print(f"    Codes: {codes}")
    print(f"    Encoded: {encoded} ({len(encoded)} bits)")
    print(f"    Original size: {len(text) * 8} bits")
    print(f"    Decoded: '{decoded}'")

    # 6. Coin Change
    print("\n[6] Coin Change")
    coins = [500, 100, 50, 10]
    amount = 1260
    result = coin_change_greedy(coins, amount)
    print(f"    Coins: {coins}, Amount: {amount}")
    print(f"    Result: {result} ({len(result)} coins)")

    # 7. Jump Game
    print("\n[7] Jump Game")
    nums1 = [2, 3, 1, 1, 4]
    nums2 = [3, 2, 1, 0, 4]
    print(f"    {nums1}: reachable = {can_jump(nums1)}, min jumps = {min_jumps(nums1)}")
    print(f"    {nums2}: reachable = {can_jump(nums2)}")

    # 8. Gas Station
    print("\n[8] Gas Station Problem")
    gas = [1, 2, 3, 4, 5]
    cost = [3, 4, 5, 1, 2]
    start = can_complete_circuit(gas, cost)
    print(f"    gas: {gas}, cost: {cost}")
    print(f"    Starting station: {start}")

    # 9. Burst Balloons
    print("\n[9] Minimum Arrows to Burst Balloons")
    points = [[10, 16], [2, 8], [1, 6], [7, 12]]
    arrows = min_arrows(points)
    print(f"    Balloons: {points}")
    print(f"    Arrows needed: {arrows}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: examples/Algorithm/c/19_greedy.c
/*
 * Greedy Algorithm
 * Activity Selection, Huffman Coding, Fractional Knapsack
 *
 * Finds the global optimum by making the locally optimal choice at each step.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* =============================================================================
 * 1. Activity Selection Problem
 * ============================================================================= */

typedef struct {
    int start;
    int end;
} Activity;

int compare_activities(const void* a, const void* b) {
    return ((Activity*)a)->end - ((Activity*)b)->end;
}

int activity_selection(Activity activities[], int n) {
    qsort(activities, n, sizeof(Activity), compare_activities);

    int count = 1;
    int last_end = activities[0].end;

    for (int i = 1; i < n; i++) {
        if (activities[i].start >= last_end) {
            count++;
            last_end = activities[i].end;
        }
    }

    return count;
}

/* =============================================================================
 * 2. Fractional Knapsack Problem
 * ============================================================================= */

typedef struct {
    double value;
    double weight;
    double ratio;
} Item;

int compare_items(const void* a, const void* b) {
    double r1 = ((Item*)a)->ratio;
    double r2 = ((Item*)b)->ratio;
    return (r2 > r1) - (r2 < r1);  /* Descending order */
}

double fractional_knapsack(Item items[], int n, double capacity) {
    qsort(items, n, sizeof(Item), compare_items);

    double total_value = 0;
    double remaining = capacity;

    for (int i = 0; i < n && remaining > 0; i++) {
        if (items[i].weight <= remaining) {
            total_value += items[i].value;
            remaining -= items[i].weight;
        } else {
            total_value += items[i].ratio * remaining;
            remaining = 0;
        }
    }

    return total_value;
}

/* =============================================================================
 * 3. Huffman Coding
 * ============================================================================= */

typedef struct HuffmanNode {
    char ch;
    int freq;
    struct HuffmanNode* left;
    struct HuffmanNode* right;
} HuffmanNode;

HuffmanNode* create_huffman_node(char ch, int freq) {
    HuffmanNode* node = malloc(sizeof(HuffmanNode));
    node->ch = ch;
    node->freq = freq;
    node->left = node->right = NULL;
    return node;
}

/* Simple priority queue (array-based) */
typedef struct {
    HuffmanNode** data;
    int size;
    int capacity;
} MinHeap;

MinHeap* create_heap(int capacity) {
    MinHeap* h = malloc(sizeof(MinHeap));
    h->data = malloc(capacity * sizeof(HuffmanNode*));
    h->size = 0;
    h->capacity = capacity;
    return h;
}

void heap_insert(MinHeap* h, HuffmanNode* node) {
    int i = h->size++;
    h->data[i] = node;
    while (i > 0 && h->data[(i - 1) / 2]->freq > h->data[i]->freq) {
        HuffmanNode* temp = h->data[(i - 1) / 2];
        h->data[(i - 1) / 2] = h->data[i];
        h->data[i] = temp;
        i = (i - 1) / 2;
    }
}

HuffmanNode* heap_extract(MinHeap* h) {
    HuffmanNode* min = h->data[0];
    h->data[0] = h->data[--h->size];
    int i = 0;
    while (2 * i + 1 < h->size) {
        int smallest = i;
        if (h->data[2 * i + 1]->freq < h->data[smallest]->freq) smallest = 2 * i + 1;
        if (2 * i + 2 < h->size && h->data[2 * i + 2]->freq < h->data[smallest]->freq) smallest = 2 * i + 2;
        if (smallest == i) break;
        HuffmanNode* temp = h->data[i];
        h->data[i] = h->data[smallest];
        h->data[smallest] = temp;
        i = smallest;
    }
    return min;
}

HuffmanNode* build_huffman_tree(char chars[], int freqs[], int n) {
    MinHeap* heap = create_heap(n);
    for (int i = 0; i < n; i++) {
        heap_insert(heap, create_huffman_node(chars[i], freqs[i]));
    }

    while (heap->size > 1) {
        HuffmanNode* left = heap_extract(heap);
        HuffmanNode* right = heap_extract(heap);
        HuffmanNode* merged = create_huffman_node('\0', left->freq + right->freq);
        merged->left = left;
        merged->right = right;
        heap_insert(heap, merged);
    }

    HuffmanNode* root = heap_extract(heap);
    free(heap->data);
    free(heap);
    return root;
}

void print_huffman_codes(HuffmanNode* root, char code[], int depth) {
    if (!root) return;
    if (!root->left && !root->right) {
        code[depth] = '\0';
        printf("      '%c': %s\n", root->ch, code);
        return;
    }
    code[depth] = '0';
    print_huffman_codes(root->left, code, depth + 1);
    code[depth] = '1';
    print_huffman_codes(root->right, code, depth + 1);
}

void free_huffman_tree(HuffmanNode* root) {
    if (!root) return;
    free_huffman_tree(root->left);
    free_huffman_tree(root->right);
    free(root);
}

/* =============================================================================
 * 4. Coin Change (Greedy)
 * ============================================================================= */

int coin_change_greedy(int coins[], int n, int amount) {
    int count = 0;
    for (int i = n - 1; i >= 0 && amount > 0; i--) {
        count += amount / coins[i];
        amount %= coins[i];
    }
    return (amount == 0) ? count : -1;
}

/* =============================================================================
 * 5. Meeting Room Assignment (Minimum Rooms)
 * ============================================================================= */

typedef struct {
    int time;
    int type;  /* 1: start, -1: end */
} Event;

int compare_events(const void* a, const void* b) {
    Event* e1 = (Event*)a;
    Event* e2 = (Event*)b;
    if (e1->time != e2->time) return e1->time - e2->time;
    return e1->type - e2->type;  /* End events first */
}

int min_meeting_rooms(int starts[], int ends[], int n) {
    Event* events = malloc(2 * n * sizeof(Event));
    for (int i = 0; i < n; i++) {
        events[2 * i] = (Event){starts[i], 1};
        events[2 * i + 1] = (Event){ends[i], -1};
    }
    qsort(events, 2 * n, sizeof(Event), compare_events);

    int max_rooms = 0, current = 0;
    for (int i = 0; i < 2 * n; i++) {
        current += events[i].type;
        if (current > max_rooms) max_rooms = current;
    }
    free(events);
    return max_rooms;
}

/* =============================================================================
 * 6. Jump Game
 * ============================================================================= */

int can_jump(int nums[], int n) {
    int max_reach = 0;
    for (int i = 0; i < n; i++) {
        if (i > max_reach) return 0;
        if (i + nums[i] > max_reach) max_reach = i + nums[i];
    }
    return 1;
}

int min_jumps(int nums[], int n) {
    if (n <= 1) return 0;
    int jumps = 0, current_end = 0, farthest = 0;
    for (int i = 0; i < n - 1; i++) {
        if (i + nums[i] > farthest) farthest = i + nums[i];
        if (i == current_end) {
            jumps++;
            current_end = farthest;
            if (current_end >= n - 1) break;
        }
    }
    return jumps;
}

/* =============================================================================
 * Test
 * ============================================================================= */

int main(void) {
    printf("============================================================\n");
    printf("Greedy Algorithm Examples\n");
    printf("============================================================\n");

    /* 1. Activity Selection */
    printf("\n[1] Activity Selection Problem\n");
    Activity activities[] = {{1, 4}, {3, 5}, {0, 6}, {5, 7}, {3, 9}, {5, 9}};
    printf("    Max activities: %d\n", activity_selection(activities, 6));

    /* 2. Fractional Knapsack */
    printf("\n[2] Fractional Knapsack\n");
    Item items[] = {{60, 10, 6}, {100, 20, 5}, {120, 30, 4}};
    printf("    Max value for capacity 50: %.2f\n", fractional_knapsack(items, 3, 50));

    /* 3. Huffman Coding */
    printf("\n[3] Huffman Coding\n");
    char chars[] = {'a', 'b', 'c', 'd', 'e'};
    int freqs[] = {5, 9, 12, 13, 16};
    HuffmanNode* root = build_huffman_tree(chars, freqs, 5);
    printf("    Codes:\n");
    char code[100];
    print_huffman_codes(root, code, 0);
    free_huffman_tree(root);

    /* 4. Coin Change */
    printf("\n[4] Coin Change\n");
    int coins[] = {1, 5, 10, 50, 100, 500};
    printf("    Coins: [1,5,10,50,100,500], Amount: 730\n");
    printf("    Minimum coins: %d\n", coin_change_greedy(coins, 6, 730));

    /* 5. Meeting Rooms */
    printf("\n[5] Minimum Meeting Rooms\n");
    int starts[] = {0, 5, 15};
    int ends[] = {30, 10, 20};
    printf("    Meetings: (0-30), (5-10), (15-20)\n");
    printf("    Rooms needed: %d\n", min_meeting_rooms(starts, ends, 3));

    /* 6. Jump Game */
    printf("\n[6] Jump Game\n");
    int nums1[] = {2, 3, 1, 1, 4};
    int nums2[] = {3, 2, 1, 0, 4};
    printf("    [2,3,1,1,4]: reachable=%s, min_jumps=%d\n",
           can_jump(nums1, 5) ? "yes" : "no", min_jumps(nums1, 5));
    printf("    [3,2,1,0,4]: reachable=%s\n", can_jump(nums2, 5) ? "yes" : "no");

    printf("\n============================================================\n");

    return 0;
}
